Userland/Libraries/LibCrypto/BigFraction/BigFraction.cpp

```diff
@@ -0,0 +1,252 @@
+/*
+ * Copyright (c) 2022, Lucas Chollet <lucas.chollet@free.fr>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "BigFraction.h"
+#include <AK/Math.h>
+#include <AK/String.h>
+#include <AK/StringBuilder.h>
+#include <LibCrypto/NumberTheory/ModularFunctions.h>
+
+namespace Crypto {
+
+BigFraction::BigFraction(Crypto::SignedBigInteger const& numerator, Crypto::UnsignedBigInteger const& denominator)
+    : m_numerator(numerator)
+    , m_denominator(m_numerator == Crypto::SignedBigInteger { 0 } ? 1 : denominator)
+{
+    VERIFY(m_denominator != 0);
+    reduce();
+}
+
+BigFraction::BigFraction(Crypto::SignedBigInteger const& value)
+    : BigFraction(value, 1)
+{
+}
+
+BigFraction::BigFraction(StringView sv)
+{
+    String str = sv.to_string(); // TODO: Once we have a StringView equivalent for this C API, we won't need to create a copy for this anymore.
+    size_t first_index = 0;
+    char* dot_ptr;
+    auto const int_part = Crypto::SignedBigInteger::create_from(strtoll(&str[first_index], &dot_ptr, 10));
+    size_t dot_index = dot_ptr - str.characters();
+    if ((dot_index < str.length()) && (str[dot_index] == '.')) {
+        size_t second_index = dot_index + 1;
+        char* end_ptr;
+        auto const frac_part = Crypto::SignedBigInteger::create_from(strtoll(&str[second_index], &end_ptr, 10));
+        size_t end_index = end_ptr - str.characters();
+        auto const frac_length = Crypto::UnsignedBigInteger::create_from(end_index - second_index);
+        *this = BigFraction { int_part } + BigFraction { frac_part, frac_length };
+    } else {
+        *this = BigFraction { int_part };
+    }
```

Comment:
```suggestion
    auto maybe_dot_index = sv.find('.');
    auto integer_part_view = sv.substring(0, maybe_dot_index.value_or(sv.length()));
    auto fraction_part_view = maybe_dot_index.has_value() ? sv.substring(1 + *maybe_dot_index) : "0"sv;
    auto integer_part = Crypto::SignedBigInteger::from_base(10, integer_part_view);
    auto fractional_part = Crypto::SignedBigInteger::from_base(10, fraction_part_view);
    auto fraction_length = fraction_part_view.length();
    *this = BigFraction(move(integer_part)) + BigFraction(move(fractional_part), fraction_length);
```

(please note that bigintegers aren't free to copy, so move() where possible).

Reply:
I kept this TODO for later ^^
GitHub beginner here, how am I supposed to integrate this without stealing it?

There is an error in your code! My bad, I didn't see the bug before 😅

```cpp
auto fraction_length = fraction_part_view.length();
*this = BigFraction(move(integer_part)) + BigFraction(move(fractional_part), fraction_length);
```

Should be replaced by

```cpp
auto fraction_length = Crypto::UnsignedBigInteger::create_from(fraction_part_view.length());
*this = BigFraction(move(integer_part)) + BigFraction(move(fractional_part), Crypto::NumberTheory::Power("10"_bigint, fractional_length));
```

Reply:
> how am I supposed to integrate this without stealing it?

it's alright to just take it (at least, my suggestions), I don't really mind or care :P

> There is an error in your code!

very likely, I wrote that on a phone with no compilers around ^^'

it looks like I just copied your code there, this doesn't seem to be `frac_length = 10^fractional_length`, just `BigInt::from(fractional_length)`
```c++
auto const frac_length = Crypto::UnsignedBigInteger::create_from(end_index - second_index);
*this = BigFraction { int_part } + BigFraction { frac_part, frac_length };
```


anyway, feel free to take the change, I wrote it to show how it can be written, if you want to copy it, go for it : )

Reply:
Sorry I haven't been clear, I didn't compile your code either but there is a logic error that I introduced when porting to BigFraction, and I didn't see it before.
I was kidding when I say that you wrote an error, I did. But I've only seen it when you rewrite it.
I tested it, and it's definitely `frac_length = 10^fractional_length` that should be used.

Anyway, stolen, adapted and committed